### DISCUSSION TITLE
Query on a composite column and mypy

### DISCUSSION THREAD

**Discussion by slafs** [sqlalchemy/sqlalchemy#11232] [Usage Questions] [answered]
Given this snippet:
```py
from dataclasses import dataclass

from sqlalchemy import Integer, String, create_engine, select
from sqlalchemy.orm import DeclarativeBase, Mapped, composite, mapped_column, sessionmaker

db_url = "sqlite:///:memory:"

engine = create_engine(db_url, echo=True)

get_session = sessionmaker(bind=engine)


class Base(DeclarativeBase): ...


@dataclass
class Money:
    amount: int
    currency: str


class DBItem(Base):
    __tablename__ = "composite_demo"

    id: Mapped[int] = mapped_column(primary_key=True)

    price: Mapped[Money] = composite(
        mapped_column("amount", Integer),
        mapped_column("currency", String(3)),
    )


def test_check_composite_query() -> None:
    db_item = DBItem(id=1, price=Money(amount=100, currency="USD"))

    with get_session() as session:
        session.add(db_item)
        session.commit()

    stmt = select(DBItem).where(DBItem.currency == "USD")  # <--- !!!
    with get_session() as session:
        obj = session.scalar(stmt)
        if obj is None:
            raise ValueError("No item found")

    assert obj.price.currency == "USD", f"Expected USD, got {obj.price.currency}"


if __name__ == "__main__":
    Base.metadata.create_all(engine)
    test_check_composite_query()
```

which works fine at runtime, but fails the `mypy --strict` check with:
```
composite_check.py:40: error: "type[DBItem]" has no attribute "currency"  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

I realize it might be odd to directly query on a column that's only part of the composite, but I feel like in this case it makes sense 😅.

Anyway, am I doing something wrong here? Is there any way to please mypy in this case?
Thank you for a great library and thanks in advance for any pointers on this problem.

Versions used:
```
# pip freeze
greenlet==3.0.3
mypy==1.9.0
mypy-extensions==1.0.0
SQLAlchemy==2.0.29
typing_extensions==4.10.0
```

```
# python -V
Python 3.11.9
```

**Comment by zzzeek** [comment]
Ideally it would work like this:

```
stmt = select(DBItem).where(DBItem.price.currency == "USD")
```

that passes typing but the composite attribute doesn't apply this namespace right now so `.currency` is not there at runtime.

There's no dynamic mechanism for `DBItem.currency` to be present automatically in typing since no code runs, so you would have to declare this:

```py
class DBItem(Base):
    __tablename__ = "composite_demo"

    id: Mapped[int] = mapped_column(primary_key=True)

    price: Mapped[Money] = composite(
        mapped_column("amount", Integer),
        mapped_column("currency", String(3)),
    )

    if TYPE_CHECKING:
        currency: Mapped[str]

```

ideally we should enhance `composite()` to provide for those attributes.

**Reply by zzzeek** [reply]
@CaselIT we should enhance `Composite` to provide for the `.columns` on the descriptor.

**Reply by zzzeek** [reply]
Hmmmm it might not work though because typing does not see `DBItem.price.currency` as `Mapped`.  we'd need a lot more tricks

**Reply by CaselIT** [reply, marked as answer]
I think the best option that we more prominently highlight option two here https://docs.sqlalchemy.org/en/20/orm/composites.html#other-mapping-forms-for-composites as the best one to enable typing support

**Reply by CaselIT** [reply]
added https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5246

**Reply by slafs** [reply]
Thanks. That's enough for me. Although the original form with `DBItem.price.currency` that @zzzeek mentioned (that's not working on runtime ATM) would be ideal. Are there any plans to support it in the future?

**Reply by zzzeek** [reply]
I dont know that it's possible unless your dataclass itself used `Mapped` like this:

```py
@dataclass
class Money:
    amount: Mapped[int]
    currency: Mapped[str]
```

which itself would...I think works with the dataclass, since `Mapped[]` is a descriptor, but based on discussions I've had in the past with dataclass people is a little risky since it's not an official pattern

**Reply by zzzeek** [reply]
I mean it's possible at runtime, above is to get it to work fully at typing time

**Reply by slafs** [reply]
Well... sure, it's just a bit inconsistent, isn't it? In case if I define only the `composite` attribute (without the separate attributes as suggested by @CaselIT), I get different behavior at runtime and at typecheck depending on whether I'm using the `currency` attribute from a class or an instance of that class.

|                        | at runtime | at typecheck |
|------------------------|------------|--------------|
| DBItem.currency        | ✅          | ❌            |
| DBItem.price.currency  | ❌          | ✅            |
| db_item.currency       | ✅          | ❌            |
| db_item.price.currency | ✅          | ✅            |

**Reply by zzzeek** [reply]
to be clear re: the typecheck issue:

```py
@dataclass
class Money:
    amount: int
    currency: Mapped[str]


class DBItem(Base):
    __tablename__ = "composite_demo"

    id: Mapped[int] = mapped_column(primary_key=True)

    price: Mapped[Money] = composite(
        mapped_column("amount", Integer),
        mapped_column("currency", String(3)),
    )

    currency: Mapped[str]

expr1 = select(DBItem.currency)

expr2 = select(DBItem.price.currency)

expr3 = select(DBItem.price.totally_made_up_name)

reveal_type(expr1)
reveal_type(expr2)
reveal_type(expr3)

```

output:

```
test3.py:40: note: Revealed type is "sqlalchemy.sql.selectable.Select[tuple[builtins.str]]"
test3.py:41: note: Revealed type is "sqlalchemy.sql.selectable.Select[Any]"
test3.py:42: note: Revealed type is "sqlalchemy.sql.selectable.Select[Any]"
```

I dont think there's any way to re-export the namespace here, that `Any` is coming out because `InstrumentedAttribute` has a `__getattr__()` method.

**Reply by zzzeek** [reply]
only a syntax that would be very cumbersome like this could make this work, since Python typing always needs an indented list of names in order to have a typed namespace:

```py
class DBItem(Base):
    __tablename__ = "composite_demo"

    id: Mapped[int] = mapped_column(primary_key=True)

    class price(composite):
        amount: Mapped[int]
        currency: Mapped[str]

```

the above syntax does **not** work but it would need to be some kind of nested blocks syntax for it to be possible

**Reply by CaselIT** [reply]
nesting a class does not work since you always get the `InstrumentedAttribute` here:
```py
DBItem.price.amount # resolved to InstrumentedAttribute
DBItem().price.amount # resolved to InstrumentedAttribute
```

you would need to instanciate `price` to make it work
```py
DBItem().price().amount # resolved to int
```
This does not make much sense.

Maybe using something else compared to `Mapped` we could export a namespace for the class attributes, like
```py
DBItem().price.c.amount # or any other name for the namespace
```
but I'm not 100% this can be correctly typed in all cases

**Reply by zzzeek** [reply]
oh no, it wont work at all :)   just saying the only way to get a namespace `myobject.foobar.x`, `myobject.foobar.y` etc. is that there are indented namespaces somewhere setting them both up

**Reply by CaselIT** [reply]
that does work if you manually tell the typechecker that there a namespece of the specified type.

Totally not working example just as example:
```py
T = TypeVar("T")


class MappedComposite(Generic[T]):
    c: type[T]

    @overload
    def __get__(self, instance: None, owner: Any) -> "MappedComposite[T]": ...

    @overload
    def __get__(self, instance: object, owner: Any) -> T: ...

    def __get__(
        self, instance: Optional[object], owner: Any
    ) -> "MappedComposite[T] | T": ...


@dataclass
class Money:
    amount: orm.Mapped[int] # Mapped is needed
    currency: orm.Mapped[str] # Mapped is needed


class DBItem(Base):
    __tablename__ = "composite_demo"

    id: orm.Mapped[int] = orm.mapped_column(primary_key=True)

    price: MappedComposite[Money]


DBItem.price.c.amount # InstrumentedAttribute[int]
DBItem().price.amount # int
DBItem().price.c # error
```

**Reply by Alc-Alc** [reply]
> I dont know that it's possible unless your dataclass itself used `Mapped` like this:
> 
> ```python
> @dataclass
> class Money:
>     amount: Mapped[int]
>     currency: Mapped[str]
> ```
> 
> which itself would...I think works with the dataclass, since `Mapped[]` is a descriptor, but based on discussions I've had in the past with dataclass people is a little risky since it's not an official pattern

Just throwing in an idea here, feel free to ignore 🙂. Would a pattern like this be better?

```py
from typing import Annotated
from dataclasses import dataclass
from sqlalchemy.orm import mapped_column

@dataclass
class Money:
    amount: Annotated[int, mapped_column()]
    currency: Annotated[str, mapped_column()]
```

**Reply by CaselIT** [reply]
No, that would not work for type checkers, since all they would see it int/str
